### DISCUSSION TITLE
Is there a simple test for internet connectivity?

### DISCUSSION THREAD

**Discussion by kjm1102** [micropython/micropython#12121] [Using MicroPython]
I've been using sta.status('rssi') in the mistaken belief that an ip address in the return meant I have internet. However I just got ss -29 ('192.168.4.2', '255.255.255.0', '192.168.4.1', '192.168.4.1') from an AP that does NOT have internet so I'm wondering if there is something simpler than a ping that I can use to indicate internet availability?

**Comment by davefes** [comment]
This is what I use in my modified version of umail:
```py
            try:
                addr = usocket.getaddrinfo(host, port)[0][-1]
                break
            except OSError as err:
                if err.args[0] == -202: #  no network available
                    with open('errors.txt', 'a') as outfile:
                        outfile.write('no network available' + '\n')
```

**Comment by kjm1102** [comment]
Tnx Dave, looks like -202 is the one to trap.

**Comment by peterhinch** [comment]
I have used [this approach](https://github.com/peterhinch/micropython-mqtt/blob/749f04bf3df5adaa627db5a3892b31be50fa1574/mqtt_as/mqtt_as.py#L324) - sending a DNS lookup for "google.com" to 8.8.8.8 (a Google DNS server). The advantage is that the implemantation is nonblocking - `socket.getaddrinfo` can block for long periods.

**Reply by Josverl** [reply]
DNS will give you some false positives and negatives as it is meant to be cached for a defined TTL, using Google.com may further complicate matters as it uses unicast IP addresses (8.8.8.8 is actually in multiple locations,)  but for most cases you just want to know that you can reach your ISP

**Reply by peterhinch** [reply]
What is really needed is a nonblocking `.getaddrinfo`. Failing that, how best to perform a nonblocking check on basic internet connectivity?
